### ISSUE TITLE
hardcoded mac address in metal-rock64-arm64.img.xz image

### ISSUE THREAD

**Issue by jlgeering** [siderolabs/talos#7095]
## Bug Report

I have 2 Pine64 Rock64 SBC. Using both a dietPi and armbian image, I get a consistant hw / mac eth0 address:

- ce:a3:0b:... for the first one
- de:e8:02:... for the second one

but when using  metal-rock64-arm64.img.xz I get the same address for both boards (and this makes it a bit tricky to run them on the same network):

- 16:d1:15:de:70:4a for talos v1.2.9
- fe:77:61:5b:2a:3a for talos v1.3.6 and v1.3.7
- 4e:e3:dd:59:32:08 for talos v1.4.0-beta.1

interestingly, the address changed from 16:d1:15:de:70:4a to fe:77:61:5b:2a:3a when I upgraded talos from 1.2 to 1.3 (I think)

I do not have this issue with the metal-pine64-arm64.img.xz image (I also have 2 of those SBCs)

### Logs

```
❯ talosctl get link eth0 -n 10.10.99.33 -o yaml
node: 10.10.99.33
metadata:
    namespace: network
    type: LinkStatuses.net.talos.dev
    id: eth0
    version: 4
    owner: network.LinkStatusController
    phase: running
    created: 2016-01-21T08:50:36Z
    updated: 2016-01-21T08:51:14Z
spec:
    index: 8
    type: ether
    linkIndex: 0
    flags: UP,BROADCAST,RUNNING,MULTICAST,LOWER_UP
    hardwareAddr: fe:77:61:5b:2a:3a
    permanentAddr: fe:77:61:5b:2a:3a
    broadcastAddr: ff:ff:ff:ff:ff:ff
    mtu: 1500
    queueDisc: mq
    operationalState: up
    kind: ""
    slaveKind: ""
    driver: rk_gmac-dwmac
    driverVersion: Jan_2016
    linkState: true
    speedMbit: 1000
    port: TwistedPair
    duplex: Full
```

### Environment

- Talos version: 
  - v1.2.9
  - v1.3.6
  - v1.3.7
  - v1.4.0-beta.1
- Kubernetes version: v1.26.3
- Platform: [Pine64 Rock64 SBC](https://www.talos.dev/v1.3/talos-guides/install/single-board-computers/rock64/)

**Comment by jlgeering** [comment]
As a temp workaround, does someone know how to "overwrite" the mac address?

I see no way to configure it in https://www.talos.dev/v1.3/reference/configuration/#device (only ref is as device selector)

**Comment by jlgeering** [comment]
and as a sidenote ... I am also curious about the difference between hardwareAddr and permanentAddr

```
❯ talosctl get link eth0 -n 10.10.99.33 -o yaml
node: 10.10.99.33
metadata:
    namespace: network
    type: LinkStatuses.net.talos.dev
    id: eth0
    version: 4
    ...
spec:
    index: 8
    type: ether
    linkIndex: 0
    flags: UP,BROADCAST,RUNNING,MULTICAST,LOWER_UP
    hardwareAddr: fe:77:61:5b:2a:3a
    permanentAddr: fe:77:61:5b:2a:3a
    ...
```

**Comment by smira** [comment]
Hardware addr can be overridden in Linux (depends on hardware), and it might change, as the link gets bonded for example, but the permanent addr is the one which can't be changed (comes from the hardware).

I have no good explanation for this, this might be either a bug in Linux kernel or u-boot firmware for this board. Talos doesn't itself make any changes to the mac addresses.

**Comment by frezbo** [comment]
This seems to be a bug with u-boot

**Comment by frezbo** [comment]
probably needs some patches for the rock64 image

**Comment by jlgeering** [comment]
@smira is there an easy / recommended way to overridden the mac address in talos linux?

@frezbo shall I create an issue in the u-boot repo?

**Comment by frezbo** [comment]
> @frezbo shall I create an issue in the u-boot repo?

upstream u-boot already probably has somehting in their mailing lists, these things take a very long time

it's mostly about adding some patches that are not in upstream u-boot but in armbian or other distros (we do that for some sbc's), but it's hard to nail down every single patch that we need

**Comment by smira** [comment]
@jlgeering Talos has now way today to change the MAC address

**Comment by jlgeering** [comment]
@smira I'm guessing you mean no way (now => no typo?)

**Comment by frezbo** [comment]
> @smira I'm guessing you mean no way (now => no typo?)

yes, correct

**Comment by jpraczyk** [comment]
Armbian currently carries two patches to enable the generation of "stable" MAC addresses on rock64. One for the arm trusted firmware, the other one is needed for u-boot.
https://github.com/armbian/build/blob/main/patch/atf/atf-rockchip64/rk3328-efuse-init.patch
https://github.com/armbian/build/blob/main/patch/u-boot/u-boot-rockchip64/rk3328-efuse-driver.patch

**Comment by ehbello** [comment]
Hello to all,

I have discovered that the upstream branch of u-boot from version 2024-07.rc1 onwards applies patches similar to those mentioned by @jpraczyk. Specifically these:

https://source.denx.de/u-boot/u-boot/-/commit/19b4caf32175f227868e9bc3185207f53b9d651f
https://source.denx.de/u-boot/u-boot/-/commit/7d4c773e66f37729d8a429fe25678b9a24094b1d

I have created the corresponding Pull Request (https://github.com/siderolabs/sbc-rockchip/pull/15) in sbc-rockchip repository to apply the changes to the main branch.

Regards.
